datafusion/datasource/src/nested_schema_adapter.rs

```diff
@@ -0,0 +1,943 @@
+//! physical format into how they should be used by DataFusion.  For instance, a schema
+//! can be stored external to a parquet file that maps parquet logical types to arrow types.
+
+use arrow::datatypes::{DataType::Struct, Field, Schema, SchemaRef};
+use datafusion_common::{ColumnStatistics, Result};
+use std::sync::Arc;
+
+use crate::schema_adapter::{
+    create_field_mapping, DefaultSchemaAdapterFactory, SchemaAdapter,
+    SchemaAdapterFactory, SchemaMapper,
+};
+use arrow::array::{Array, ArrayRef, StructArray};
+use arrow::compute::cast;
+use arrow::record_batch::{RecordBatch, RecordBatchOptions};
+use datafusion_common::arrow::array::new_null_array;
+
+/// Factory for creating [`NestedStructSchemaAdapter`]
+///
+/// This factory creates schema adapters that properly handle schema evolution
+/// for nested struct fields, allowing new fields to be added to struct columns
+/// over time.
+#[derive(Debug, Clone, Default)]
+pub struct NestedStructSchemaAdapterFactory;
+
+impl SchemaAdapterFactory for NestedStructSchemaAdapterFactory {
+    fn create(
+        &self,
+        projected_table_schema: SchemaRef,
+        table_schema: SchemaRef,
+    ) -> Box<dyn SchemaAdapter> {
+        Box::new(NestedStructSchemaAdapter::new(
+            projected_table_schema,
+            table_schema,
+        ))
+    }
+}
+
+impl NestedStructSchemaAdapterFactory {
+    /// Create a new factory for mapping batches from a file schema to a table
+    /// schema with support for nested struct evolution.
+    ///
+    /// This is a convenience method that handles nested struct fields properly.
+    pub fn from_schema(table_schema: SchemaRef) -> Box<dyn SchemaAdapter> {
+        Self.create(Arc::clone(&table_schema), table_schema)
+    }
+
+    /// Determines if a schema contains nested struct fields that would benefit
+    /// from special handling during schema evolution
+    pub fn has_nested_structs(schema: &Schema) -> bool {
+        schema
+            .fields()
+            .iter()
+            .any(|field| matches!(field.data_type(), Struct(_)))
+    }
+
+    /// Create an appropriate schema adapter based on schema characteristics.
+    /// Returns a NestedStructSchemaAdapter if the projected schema contains nested structs,
+    /// otherwise returns a DefaultSchemaAdapter.
+    pub fn create_adapter(
+        projected_table_schema: SchemaRef,
+        table_schema: SchemaRef,
+    ) -> Box<dyn SchemaAdapter> {
+        // Use nested adapter if target has nested structs
+        if Self::has_nested_structs(table_schema.as_ref()) {
+            NestedStructSchemaAdapterFactory.create(projected_table_schema, table_schema)
+        } else {
+            // Default case for simple schemas
+            DefaultSchemaAdapterFactory.create(projected_table_schema, table_schema)
+        }
+    }
+}
+
+/// A SchemaAdapter that handles schema evolution for nested struct types
+#[derive(Debug, Clone)]
+pub struct NestedStructSchemaAdapter {
```

Comment:
People have requested something like this for other DataFusion operations (such as `CAST` and coercing structs to other types) -- I wonder if there is some way we make the logic more reusable 🤔 

Something like separate the schema mapping structure and actual logic out of the datasource crate